File: scripts/dq-check.mjs
/**
 * Data quality check across all WDR files.
 * Flags physically impossible or highly suspicious T/S parameter values.
 *
 * Run: node scripts/dq-check.mjs [--collection <name>] [--check-urls]
 *
 * --check-urls  HTTP HEAD every boxbench_datasheet / boxbench_manu_page URL
 *               in flagged files and report status + content-type.
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const DRIVERS_DIR = path.join(path.dirname(fileURLToPath(import.meta.url)), '..', 'drivers');
const _ci = process.argv.indexOf('--collection');
const filterColl = _ci >= 0 ? process.argv[_ci + 1] : null;
const checkUrls  = process.argv.includes('--check-urls');

function parseFields(text) {
  const f = {};
  for (const line of text.split(/\r?\n/)) {
    const eq = line.indexOf('=');
    if (eq < 0 || line[0] === '[') continue;
    f[line.slice(0, eq).trim()] = line.slice(eq + 1).trim();
  }
  return f;
}
const n = (f, k) => { const v = parseFloat(f[k]); return isFinite(v) ? v : null; };

// ── DQ rules ─────────────────────────────────────────────────────────────────
// Each rule: { id, desc, test(fields) → string|null }
// test returns null (pass) or a short description of what's wrong.
//
// Zero-value policy: real datasheets never publish 0 for a T/S parameter.
// A 0 in Sd, Re, BL, Mms, Pe, Xmax, Qts, Qms, Vas is ALWAYS a scraper
// artifact (blank cell → 0). These are flagged as "zero = scraper artifact".
const RULES = [
  // ── Absent-or-zero fields ─────────────────────────────────────────────────
  // "absent" = field not in file; "zero" = field present but = 0.
  // Both mean "scraper had no value". Zero is never a valid T/S parameter.
  // absent = key not in file; zero/empty = key present but no valid positive value.
  // Both are scraper artifacts — real datasheets never publish 0 for these fields.
  { id: 'missing_Fs',  desc: 'Fs absent, zero, or invalid — scraper artifact',
    test: f => { const v = n(f,'Fs');
      if (!('Fs' in f))              return 'Fs absent';
      if (v === 0)                   return 'Fs=0 (scraper artifact)';
      if (v === null || v < 0)       return `Fs=${f.Fs} (invalid)`;
      return null; } },
  { id: 'missing_Sd',  desc: 'Sd absent, zero, or invalid — scraper artifact',
    test: f => { const v = n(f,'Sd');
      if (!('Sd' in f))              return 'Sd absent';
      if (v === 0)                   return 'Sd=0 (scraper artifact)';
      if (v === null || v < 0)       return `Sd=${f.Sd} (invalid)`;
      return null; } },
  { id: 'missing_Re',  desc: 'Re absent, zero, or invalid — scraper artifact',
    test: f => { const v = n(f,'Re');
      if (!('Re' in f))              return 'Re absent';
      if (v === 0)                   return 'Re=0 (scraper artifact)';
      if (v === null || v < 0)       return `Re=${f.Re} (invalid)`;
      return null; } },
  { id: 'zero_BL',  desc: 'BL = 0 — scraper artifact; no real motor has zero force factor',
    test: f => ('BL'  in f && n(f,'BL')  === 0) ? 'BL=0'  : null },
  { id: 'zero_Mms', desc: 'Mms = 0 — scraper artifact; no cone has zero moving mass',
    test: f => ('Mms' in f && n(f,'Mms') === 0) ? 'Mms=0' : null },
  { id: 'zero_Qts', desc: 'Qts = 0 — scraper artifact; Qts=0 is thermodynamically impossible',
    test: f => ('Qts' in f && n(f,'Qts') === 0) ? 'Qts=0' : null },
  { id: 'zero_Qms', desc: 'Qms = 0 — scraper artifact; no suspension has zero mechanical Q',
    test: f => ('Qms' in f && n(f,'Qms') === 0) ? 'Qms=0' : null },
  { id: 'zero_Vas', desc: 'Vas = 0 — scraper artifact; equivalent air volume cannot be zero',
    test: f => ('Vas' in f && n(f,'Vas') === 0) ? 'Vas=0' : null },

  // ── Fs range ──────────────────────────────────────────────────────────────
  // Fs_low also catches the European dot-thousands scraper bug:
  // SI page shows "1.600 Hz" (dot = thousands sep) → parse_number sees 1.600 → 1.6.
  // Same root cause as Pe=1 from "1.000 W". Fix: multiply Fs by 1000.
  { id: 'Fs_low',  desc: 'Fs < 5 Hz — physically impossible; also catches dot-thousands bug ("1.600 Hz" → 1.6)',
    test: f => { const v = n(f,'Fs'); return v && v > 0 && v < 5 ? `Fs=${v}` : null; } },
  { id: 'Fs_high', desc: 'Fs > 5000 Hz — implausible for a cone driver',
    test: f => { const v = n(f,'Fs'); return v && v > 5000 ? `Fs=${v}` : null; } },

  // ── Sd range ──────────────────────────────────────────────────────────────
  { id: 'Sd_huge', desc: 'Sd > 3000 cm² — larger than any real driver (21" sub ≈ 1700 cm², 24" ≈ 2300 cm²)',
    test: f => { const v = n(f,'Sd'); return v && v*1e4 > 3000 ? `Sd=${(v*1e4).toFixed(0)} cm²` : null; } },
  // Sd_tiny also catches the Beyma Sd=0.0001 pattern: scraper wrote a near-zero
  // instead of omitting the field. Treat same as Sd=0.
  { id: 'Sd_tiny', desc: 'Sd < 0.5 cm² — scraper artifact (near-zero); no real driver has Sd this small',
    test: f => { const v = n(f,'Sd'); return v && v*1e4 < 0.5 ? `Sd=${(v*1e4).toFixed(3)} cm²` : null; } },

  // ── Re range ──────────────────────────────────────────────────────────────
  { id: 'Re_low',  desc: 'Re < 1 Ω — below DC resistance of any voice coil',
    test: f => { const v = n(f,'Re'); return v && v < 1   ? `Re=${v}` : null; } },
  { id: 'Re_high', desc: 'Re > 64 Ω — implausibly high voice coil resistance',
    test: f => { const v = n(f,'Re'); return v && v > 64  ? `Re=${v}` : null; } },

  // ── Q values ──────────────────────────────────────────────────────────────
  { id: 'Qts_impossible', desc: 'Qts > Qes — thermodynamically impossible (Qts must be < Qes)',
    test: f => { const qts = n(f,'Qts'), qes = n(f,'Qes'); return qts && qes && qts >= qes ? `Qts=${qts} Qes=${qes}` : null; } },
  { id: 'Qts_impossible2', desc: 'Qts > Qms — thermodynamically impossible',
    test: f => { const qts = n(f,'Qts'), qms = n(f,'Qms'); return qts && qms && qts >= qms ? `Qts=${qts} Qms=${qms}` : null; } },
  { id: 'Qts_high', desc: 'Qts > 5 — physically unreasonable for any driver',
    test: f => { const v = n(f,'Qts'); return v && v > 5  ? `Qts=${v}` : null; } },
  { id: 'Qes_zero', desc: 'Qes ≤ 0 — impossible (zero electrical Q = infinite motor damping)',
    test: f => { const v = n(f,'Qes'); return v !== null && v <= 0 ? `Qes=${v}` : null; } },
  { id: 'Qms_low', desc: 'Qms < 0.5 — extremely lossy suspension, very unusual',
    test: f => { const v = n(f,'Qms'); return v && v < 0.5 ? `Qms=${v}` : null; } },

  // ── Pe ────────────────────────────────────────────────────────────────────
  // Pe_one: "1,000 W" or "1.000 W" → scraper dot/comma bug → stored as 1.
  // Pe_zero: blank cell → scraper stored 0 instead of omitting field.
  { id: 'Pe_one',  desc: 'Pe = 1 W — dot/comma-thousands scraper bug (e.g. "1.000 W" or "1,000 W" → 1)',
    test: f => { const v = n(f,'Pe'); return v === 1 ? 'Pe=1 (scraper artifact)' : null; } },
  { id: 'Pe_zero', desc: 'Pe = 0 — scraper artifact; no datasheet publishes Pe=0',
    test: f => { const v = n(f,'Pe'); return v !== null && v === 0 ? 'Pe=0 (scraper artifact)' : null; } },

  // ── Xmax ──────────────────────────────────────────────────────────────────
  // Xmax_zero: scraper artifact (blank cell → 0).
  // Xmax_huge: mm stored as m by scraper (value should be divided by 1000).
  { id: 'Xmax_zero', desc: 'Xmax = 0 — scraper artifact; no datasheet publishes Xmax=0',
    test: f => { const v = n(f,'Xmax'); return v !== null && v === 0 ? 'Xmax=0 (scraper artifact)' : null; } },
  { id: 'Xmax_huge', desc: 'Xmax > 100 mm — mm stored as m by scraper; divide by 1000',
    test: f => { const v = n(f,'Xmax'); return v && v*1000 > 100 ? `Xmax=${(v*1000).toFixed(0)} mm` : null; } },

  // ── Vas ───────────────────────────────────────────────────────────────────
  { id: 'Vas_huge', desc: 'Vas > 2000 L — implausible (would need a room-sized box)',
    test: f => { const v = n(f,'Vas'); return v && v*1000 > 2000 ? `Vas=${(v*1000).toFixed(0)} L` : null; } },
  // ft³-as-liters scraper bug: scraper encounters a value in ft³ on soundimports.eu,
  // divides by 1000 assuming liters → value is ~28x too small.
  // Heuristic: flag when Vas(L) < Sd(cm²) / 60 AND Sd > 150 cm².
  // Catches WO24P-4 (Sd=255, Vas=2.5L→flag at <4.25L) and SEAS L26 (Sd=507, Vas=6L→flag at <8.45L)
  // without hitting legitimate stiff PA midranges (Sd=150, threshold=2.5L, real Vas=3L → pass).
  { id: 'Vas_tiny_for_driver', desc: 'Vas implausibly small for piston area — likely ft³-as-liters scraper bug',
    test: f => {
      const vas = n(f,'Vas'), sd = n(f,'Sd');
      if (!sd || !vas) return null;
      const sd_cm2 = sd * 1e4, vas_L = vas * 1000;
      return sd_cm2 > 150 && vas_L < sd_cm2 / 60
        ? `Sd=${sd_cm2.toFixed(0)} cm² Vas=${vas_L.toFixed(2)} L (min plausible ≈ ${(sd_cm2/60).toFixed(1)} L)` : null;
    }},

  // Sd vs Fs cross-check: a tiny Sd with very low Fs is suspicious
  { id: 'tweeter_fs', desc: 'Sd < 5 cm² but Fs < 100 Hz — tiny piston with woofer-range Fs',
    test: f => {
      const sd = n(f,'Sd'), fs = n(f,'Fs');
      return sd && fs && sd*1e4 < 5 && fs < 100 ? `Sd=${(sd*1e4).toFixed(1)} cm² Fs=${fs}` : null;
    }},

  // SPL sanity
  { id: 'SPL_high', desc: 'SPL > 120 dB/1W/1m — physically implausible for a passive driver',
    test: f => { const v = n(f,'SPL'); return v && v > 120 ? `SPL=${v}` : null; } },
  { id: 'SPL_low',  desc: 'SPL < 65 dB/1W/1m — implausibly inefficient',
    test: f => { const v = n(f,'SPL'); return v && v < 65  ? `SPL=${v}` : null; } },
];

// ── URL liveness check ────────────────────────────────────────────────────────
const URL_FIELDS = ['boxbench_datasheet', 'boxbench_manu_page', 'boxbench_vendor_page'];

async function checkUrl(url) {
  try {
    const ctrl = new AbortController();
    const tid  = setTimeout(() => ctrl.abort(), 8000);
    const res  = await fetch(url, { method: 'HEAD', signal: ctrl.signal, redirect: 'follow' });
    clearTimeout(tid);
    const ct = res.headers.get('content-type') ?? 'unknown';
    const kb = res.headers.get('content-length') ? Math.round(+res.headers.get('content-length')/1024)+'KB' : '';
    return `${res.status} ${ct.split(';')[0]}${kb ? ' '+kb : ''}`;
  } catch (e) {
    return `ERR ${e.message.slice(0,60)}`;
  }
}

// ── Scan all collections ──────────────────────────────────────────────────────
const issues = []; // { collection, fname, ruleId, desc, detail, fields }

for (const coll of fs.readdirSync(DRIVERS_DIR).sort()) {
  if (filterColl && coll !== filterColl) continue;
  const collPath = path.join(DRIVERS_DIR, coll);
  if (!fs.statSync(collPath).isDirectory()) continue;
  for (const fname of fs.readdirSync(collPath).sort()) {
    if (!fname.endsWith('.wdr')) continue;
    const text = fs.readFileSync(path.join(collPath, fname), 'utf8');
    const f    = parseFields(text);
    for (const rule of RULES) {
      const hit = rule.test(f);
      if (hit) issues.push({ collection: coll, fname, ruleId: rule.id, desc: rule.desc, detail: hit, fields: f });
    }
  }
}

// ── Collect unique URLs from flagged files for liveness check ─────────────────
let urlStatus = {};
if (checkUrls) {
  const urlSet = new Set();
  for (const iss of issues) {
    for (const k of URL_FIELDS) {
      const v = iss.fields[k];
      if (v && v.startsWith('http')) urlSet.add(v);
    }
  }
  const urls = [...urlSet];
  console.log(`\nChecking ${urls.length} URLs (HEAD requests, 8 s timeout)…\n`);
  // Sequential to avoid hammering servers
  for (const url of urls) {
    process.stdout.write(`  ${url}  →  `);
    const s = await checkUrl(url);
    urlStatus[url] = s;
    console.log(s);
  }
}

// ── Report grouped by rule ────────────────────────────────────────────────────
const byRule = {};
for (const iss of issues) {
  if (!byRule[iss.ruleId]) byRule[iss.ruleId] = { desc: iss.desc, hits: [] };
  byRule[iss.ruleId].hits.push(iss);
}

let total = 0;
for (const [ruleId, { desc, hits }] of Object.entries(byRule).sort()) {
  console.log(`\n── ${ruleId} (${hits.length}) — ${desc}`);
  for (const h of hits) {
    console.log(`   ${h.collection}/${h.fname}  [${h.detail}]`);
    // Print boxbench link fields so the AI doesn't have to read the file
    for (const k of URL_FIELDS) {
      const url = h.fields[k];
      if (!url) continue;
      const status = urlStatus[url] ? `  → ${urlStatus[url]}` : '';
      console.log(`     ${k}: ${url}${status}`);
    }
    if (h.fields.boxbench_corrections) {
      console.log(`     boxbench_corrections: ${h.fields.boxbench_corrections.slice(0,120)}${h.fields.boxbench_corrections.length > 120 ? '…' : ''}`);
    }
  }
  total += hits.length;
}
console.log(`\nTotal issues: ${total} across ${issues.map(i=>i.collection+'/'+i.fname).filter((v,i,a)=>a.indexOf(v)===i).length} files`);

File: scripts/dq_check.py
"""
dq_check.py — WDR data quality rules, shared by the CLI tool and scrapers.

check_fields(fields) is the single source of truth for all DQ rules.
It is imported by scraper_lib.py to validate at write time.

Run: python scripts/dq_check.py [--collection <name>] [--check-urls]
  --collection   only check one collection subdirectory
  --check-urls   GET (Range: bytes=0-3) every URL in _meta.yml sidecars of flagged files; report real content-type
"""

import math
import pathlib
import re
import sys
import urllib.request
import urllib.error
import argparse

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

try:
    import yaml as _yaml
    _YAML_OK = True
except ImportError:
    _YAML_OK = False

DRIVERS_DIR = pathlib.Path(__file__).parent.parent / "drivers"
# URL fields live in _meta.yml sidecars (NOT in WDR files)
SIDECAR_URL_FIELDS = ["datasheet", "manu_page", "vendor_page", "frd", "impedance"]


def parse_fields(text: str) -> dict:
    fields = {}
    for line in text.splitlines():
        eq = line.find("=")
        if eq < 0 or line.startswith("["):
            continue
        fields[line[:eq].strip()] = line[eq + 1:].strip()
    return fields


def _n(f: dict, key: str):
    try:
        v = float(f.get(key, ""))
        return v if math.isfinite(v) else None
    except (TypeError, ValueError):
        return None


# ── Rules ─────────────────────────────────────────────────────────────────────
# Each rule is a plain function: fields -> str | None
# Returns None (pass) or a short detail string describing what is wrong.
# RULES list at the bottom maps each function to its id and description.

def _missing_Fs(f):
    v = _n(f, "Fs")
    if "Fs" not in f:           return "Fs absent"
    if v == 0:                  return "Fs=0 (scraper artifact)"
    if v is None or v < 0:     return f"Fs={f['Fs']} (invalid)"

def _missing_Sd(f):
    v = _n(f, "Sd")
    if "Sd" not in f:           return "Sd absent"
    if v == 0:                  return "Sd=0 (scraper artifact)"
    if v is None or v < 0:     return f"Sd={f['Sd']} (invalid)"

def _missing_Re(f):
    # Passive radiators have no voice coil — Re and BL are both absent by design.
    if "Re" not in f and "BL" not in f:
        return None
    v = _n(f, "Re")
    if "Re" not in f:           return "Re absent"
    if v == 0:                  return "Re=0 (scraper artifact)"
    if v is None or v < 0:     return f"Re={f['Re']} (invalid)"

def _zero_BL(f):
    return "BL=0" if ("BL" in f and _n(f, "BL") == 0) else None

def _zero_Mms(f):
    return "Mms=0" if ("Mms" in f and _n(f, "Mms") == 0) else None

def _zero_Qts(f):
    return "Qts=0" if ("Qts" in f and _n(f, "Qts") == 0) else None

def _zero_Qms(f):
    return "Qms=0" if ("Qms" in f and _n(f, "Qms") == 0) else None

def _zero_Vas(f):
    return "Vas=0" if ("Vas" in f and _n(f, "Vas") == 0) else None

def _Fs_low(f):
    v = _n(f, "Fs")
    return f"Fs={v}" if (v and 0 < v < 5) else None

def _Fs_high(f):
    v = _n(f, "Fs")
    return f"Fs={v}" if (v and v > 5000) else None

def _Fs_no_Q(f):
    v = _n(f, "Fs")
    if not v or v <= 500:
        return None
    if "Qts" not in f and "Qes" not in f:
        return (f"Fs={v:.0f} Hz with no Qts/Qes — likely crossover/range freq, "
                f"not T/S resonance; compression drivers and AMTs don't publish T/S Fs")

def _Sd_huge(f):
    v = _n(f, "Sd")
    return f"Sd={v*1e4:.0f} cm²" if (v and v * 1e4 > 3000) else None

def _Sd_tiny(f):
    v = _n(f, "Sd")
    return f"Sd={v*1e4:.3f} cm²" if (v and 0 < v * 1e4 < 0.5) else None

def _Re_low(f):
    v = _n(f, "Re")
    return f"Re={v}" if (v and v < 1) else None

def _Re_high(f):
    v = _n(f, "Re")
    return f"Re={v}" if (v and v > 64) else None

def _Qts_impossible(f):
    qts, qes = _n(f, "Qts"), _n(f, "Qes")
    return f"Qts={qts} Qes={qes}" if (qts and qes and qts >= qes) else None

def _Qts_impossible2(f):
    qts, qms = _n(f, "Qts"), _n(f, "Qms")
    return f"Qts={qts} Qms={qms}" if (qts and qms and qts >= qms) else None

def _Qts_high(f):
    v = _n(f, "Qts")
    return f"Qts={v}" if (v and v > 5) else None

def _Qes_zero(f):
    v = _n(f, "Qes")
    return f"Qes={v}" if (v is not None and v <= 0) else None

def _Qms_low(f):
    v = _n(f, "Qms")
    return f"Qms={v}" if (v and v < 0.5) else None

def _Pe_one(f):
    v = _n(f, "Pe")
    return "Pe=1 (scraper artifact)" if v == 1 else None

def _Pe_zero(f):
    v = _n(f, "Pe")
    return "Pe=0 (scraper artifact)" if (v is not None and v == 0) else None

def _Xmax_zero(f):
    v = _n(f, "Xmax")
    return "Xmax=0 (scraper artifact)" if (v is not None and v == 0) else None

def _Xmax_huge(f):
    v = _n(f, "Xmax")
    return f"Xmax={v*1000:.0f} mm" if (v and v * 1000 > 100) else None

def _Vas_huge(f):
    v = _n(f, "Vas")
    return f"Vas={v*1000:.0f} L" if (v and v * 1000 > 2000) else None

def _Vas_tiny_for_driver(f):
    vas, sd = _n(f, "Vas"), _n(f, "Sd")
    if not sd or not vas:
        return None
    sd_cm2, vas_L = sd * 1e4, vas * 1000
    if sd_cm2 > 150 and vas_L < sd_cm2 / 60:
        return (f"Sd={sd_cm2:.0f} cm² Vas={vas_L:.2f} L "
                f"(min plausible ≈ {sd_cm2/60:.1f} L)")

def _tweeter_fs(f):
    sd, fs = _n(f, "Sd"), _n(f, "Fs")
    if sd and fs and sd * 1e4 < 5 and fs < 100:
        return f"Sd={sd*1e4:.1f} cm² Fs={fs}"

def _SPL_high(f):
    v = _n(f, "SPL")
    return f"SPL={v}" if (v and v > 120) else None

def _SPL_low(f):
    v = _n(f, "SPL")
    return f"SPL={v}" if (v and v < 65) else None


RULES = [
    # ── Absent or zero fields ─────────────────────────────────────────────────
    ("missing_Fs",  "Fs absent, zero, or invalid — scraper artifact",                       _missing_Fs),
    ("missing_Sd",  "Sd absent, zero, or invalid — scraper artifact",                       _missing_Sd),
    ("missing_Re",  "Re absent, zero, or invalid — scraper artifact",                       _missing_Re),
    ("zero_BL",     "BL = 0 — scraper artifact; no real motor has zero force factor",       _zero_BL),
    ("zero_Mms",    "Mms = 0 — scraper artifact; no cone has zero moving mass",             _zero_Mms),
    ("zero_Qts",    "Qts = 0 — scraper artifact; Qts=0 is thermodynamically impossible",   _zero_Qts),
    ("zero_Qms",    "Qms = 0 — scraper artifact; no suspension has zero mechanical Q",      _zero_Qms),
    ("zero_Vas",    "Vas = 0 — scraper artifact; equivalent air volume cannot be zero",     _zero_Vas),
    # ── Fs ───────────────────────────────────────────────────────────────────
    ("Fs_low",    'Fs < 5 Hz — impossible; dot-thousands bug ("1.600 Hz" → 1.6); fix: ×1000', _Fs_low),
    ("Fs_high",   "Fs > 5000 Hz — implausible for a cone driver",                          _Fs_high),
    ("Fs_no_Q",   "Fs > 500 Hz but no Qts/Qes — crossover/range freq captured as Fs (compression driver or AMT)", _Fs_no_Q),
    # ── Sd ───────────────────────────────────────────────────────────────────
    ("Sd_huge",   "Sd > 3000 cm² — larger than any real driver",                           _Sd_huge),
    ("Sd_tiny",   "Sd < 0.5 cm² — scraper artifact (near-zero)",                           _Sd_tiny),
    # ── Re ───────────────────────────────────────────────────────────────────
    ("Re_low",    "Re < 1 Ω — below DC resistance of any voice coil",                      _Re_low),
    ("Re_high",   "Re > 64 Ω — implausibly high voice coil resistance",                    _Re_high),
    # ── Q values ─────────────────────────────────────────────────────────────
    ("Qts_impossible",  "Qts ≥ Qes — thermodynamically impossible (Qts must be < Qes)",   _Qts_impossible),
    ("Qts_impossible2", "Qts ≥ Qms — thermodynamically impossible",                        _Qts_impossible2),
    ("Qts_high",        "Qts > 5 — physically unreasonable for any driver",                _Qts_high),
    ("Qes_zero",        "Qes ≤ 0 — impossible (zero electrical Q = infinite motor damping)", _Qes_zero),
    ("Qms_low",         "Qms < 0.5 — extremely lossy suspension, very unusual",            _Qms_low),
    # ── Pe ───────────────────────────────────────────────────────────────────
    ("Pe_one",    'Pe = 1 W — dot/comma-thousands scraper bug ("1.000 W" → 1)',            _Pe_one),
    ("Pe_zero",   "Pe = 0 — scraper artifact; no datasheet publishes Pe=0",                _Pe_zero),
    # ── Xmax ─────────────────────────────────────────────────────────────────
    ("Xmax_zero", "Xmax = 0 — scraper artifact; no datasheet publishes Xmax=0",           _Xmax_zero),
    ("Xmax_huge", "Xmax > 100 mm — mm stored as m by scraper; divide by 1000",            _Xmax_huge),
    # ── Vas ──────────────────────────────────────────────────────────────────
    ("Vas_huge",           "Vas > 2000 L — implausible (would need a room-sized box)",    _Vas_huge),
    ("Vas_tiny_for_driver","Vas implausibly small for piston area — likely ft³-as-liters scraper bug", _Vas_tiny_for_driver),
    # ── Cross-field ──────────────────────────────────────────────────────────
    ("tweeter_fs","Sd < 5 cm² but Fs < 100 Hz — tiny piston with woofer-range Fs",       _tweeter_fs),
    # ── SPL ──────────────────────────────────────────────────────────────────
    ("SPL_high",  "SPL > 120 dB/1W/1m — physically implausible for a passive driver",    _SPL_high),
    ("SPL_low",   "SPL < 65 dB/1W/1m — implausibly inefficient",                         _SPL_low),
]


def check_fields(fields: dict) -> list[tuple[str, str, str]]:
    """
    Run all DQ rules against a parsed WDR fields dict.
    Returns list of (rule_id, description, detail) for every failing rule.
    Empty list means no issues found.
    """
    hits = []
    for rule_id, desc, fn in RULES:
        detail = fn(fields)
        if detail:
            hits.append((rule_id, desc, detail))
    return hits


# ── CLI ───────────────────────────────────────────────────────────────────────

def _check_url(url: str) -> str:
    """Fetch first 4 bytes via GET Range request and report Content-Type.

    HEAD is not used because some hosts (e.g. Parts Express CDN) return
    200 text/html for HEAD on any path regardless of whether the file
    exists. A Range GET reveals the real content-type and a non-empty body.
    """
    try:
        req = urllib.request.Request(url, headers={
            "User-Agent": "resonate-dq/1.0",
            "Range": "bytes=0-3",
        })
        with urllib.request.urlopen(req, timeout=10) as r:
            body = r.read(4)
            ct = r.headers.get("Content-Type", "unknown").split(";")[0]
            cl = r.headers.get("Content-Length", "") or r.headers.get("Content-Range", "")
            # Warn if server returns HTML instead of the expected binary type
            # Strip query string before checking extension
            path = url.split("?")[0].lower()
            warn = ""
            if path.endswith(".pdf") and "pdf" not in ct:
                warn = " ⚠ expected PDF"
            elif path.endswith(".zip") and "zip" not in ct:
                warn = " ⚠ expected ZIP"
            elif not body:
                warn = " ⚠ empty body"
            return f"{r.status} {ct}{warn}"
    except urllib.error.HTTPError as e:
        return f"{e.code} {e.reason}"
    except Exception as e:
        return f"ERR {str(e)[:60]}"


def main():
    ap = argparse.ArgumentParser(description="WDR data quality check")
    ap.add_argument("--collection", help="Only check this collection subdirectory")
    ap.add_argument("--check-urls", action="store_true",
                    help="GET (Range: bytes=0-3) every URL in _meta.yml sidecars; validates real content-type")
    args = ap.parse_args()

    issues = []  # (collection, fname, rule_id, desc, detail, fields, sidecar)

    for coll_path in sorted(DRIVERS_DIR.iterdir()):
        if not coll_path.is_dir():
            continue
        if args.collection and coll_path.name != args.collection:
            continue
        for wdr_path in sorted(coll_path.glob("*.wdr")):
            fields = parse_fields(wdr_path.read_text(encoding="utf-8", errors="replace"))
            # Load matching _meta.yml sidecar if present
            sidecar_path = wdr_path.with_name(wdr_path.stem + "_meta.yml")
            sidecar = {}
            if sidecar_path.exists() and _YAML_OK:
                try:
                    sidecar = _yaml.safe_load(sidecar_path.read_text(encoding="utf-8")) or {}
                except Exception:
                    pass
            for rule_id, desc, detail in check_fields(fields):
                issues.append((coll_path.name, wdr_path.name, rule_id, desc, detail, fields, sidecar))

    url_status = {}
    if args.check_urls:
        url_set = set()
        for *_, sidecar in issues:
            for k in SIDECAR_URL_FIELDS:
                v = sidecar.get(k) or ""
                if isinstance(v, str) and v.startswith("http"):
                    url_set.add(v)
        print(f"\nChecking {len(url_set)} URLs from sidecars (GET Range, 10 s timeout)…\n")
        for url in sorted(url_set):
            sys.stdout.write(f"  {url}  →  ")
            sys.stdout.flush()
            status = _check_url(url)
            url_status[url] = status
            print(status)

    by_rule: dict[str, dict] = {}
    for coll, fname, rule_id, desc, detail, fields, sidecar in issues:
        if rule_id not in by_rule:
            by_rule[rule_id] = {"desc": desc, "hits": []}
        by_rule[rule_id]["hits"].append((coll, fname, detail, fields, sidecar))

    total = 0
    seen_files = set()
    for rule_id, data in sorted(by_rule.items()):
        hits = data["hits"]
        print(f"\n── {rule_id} ({len(hits)}) — {data['desc']}")
        for coll, fname, detail, fields, sidecar in hits:
            print(f"   {coll}/{fname}  [{detail}]")
            for k in SIDECAR_URL_FIELDS:
                url = sidecar.get(k) or ""
                if not isinstance(url, str) or not url.startswith("http"):
                    continue
                status = f"  → {url_status[url]}" if url in url_status else ""
                print(f"     {k}: {url}{status}")
            corr = sidecar.get("corrections") or ""
            if corr:
                print(f"     corrections: {str(corr)[:120]}{'…' if len(str(corr)) > 120 else ''}")
            seen_files.add(f"{coll}/{fname}")
        total += len(hits)

    print(f"\nTotal issues: {total} across {len(seen_files)} files")


if __name__ == "__main__":
    main()
